spring-kafka/README.md

```diff
@@ -0,0 +1,357 @@
+Karaf is not following the same algorithm as Java.
+Because of that methods which accept Generic types don't recognize the existing methods as valid ones to use in the injection.
+
+Further info can be seen in the `osgi-blueprint-api` module class [GenericsTypeConverter.java](https://github.com/pentaho/pentaho-osgi-bundles/blob/54c4e1d0ad1324962714d5c405318331b4f223f4/osgi-blueprint-api/src/main/java/org/hitachivantara/osgi/service/blueprint/GenericsTypeConverter.java)
+javadoc.
+
+See also below issues in apache Aries project to more details:
+[Apache ARIES-1500](https://issues.apache.org/jira/browse/ARIES-1500)
+[Apache ARIES-1607](https://issues.apache.org/jira/browse/ARIES-1607)
+
+## Producer
+
+To produce messages to Kafka topics the following blueprint definitions should be used:
+
+```xml
+<type-converters>
+  <bean class="org.hitachivantara.osgi.service.blueprint.GenericsTypeConverter">
+    <argument>
+      <map key-type="java.lang.Class" value-type="java.lang.Class">
+        <entry key="org.apache.kafka.common.serialization.StringSerializer" value="org.apache.kafka.common.serialization.Serializer" />
+        <entry key="org.hitachivantara.spring.kafka.protobuf.KafkaProtobufSerializer" value="org.apache.kafka.common.serialization.Serializer" />
+        <entry key="org.springframework.kafka.core.DefaultKafkaProducerFactory" value="org.springframework.kafka.core.ProducerFactory" />
+
+        <entry key="org.springframework.kafka.core.KafkaTemplate" value="org.springframework.kafka.core.KafkaTemplate" />
+      </map>
+    </argument>
+  </bean>
+</type-converters>
+
+<!-- Producer properties -->
+<bean id="kafkaProducerProperties" class="java.util.HashMap">
+  <argument>
+    <map key-type="java.lang.String" value-type="java.lang.Object">
+      <entry key="bootstrap.servers" value="localhost:9092"/>
+      <entry key="retries" value="0"/>
+      <entry key="batch.size" value="16384"/>
+      <entry key="linger.ms" value="1"/>
+      <entry key="buffer.memory" value="33554432"/>
+    </map>
+  </argument>
+</bean>
+
+<!-- Producer Kafka Template -->
+<bean id="kafkaProxyProducerKafkaTemplate" class="org.springframework.kafka.core.KafkaTemplate">
+  <argument> <!-- type converter used on this argument -->
+    <bean class="org.springframework.kafka.core.DefaultKafkaProducerFactory">
+      <argument ref="kafkaProducerProperties"/>
+      <argument> <!-- type converter used on this argument -->
+        <bean class="org.apache.kafka.common.serialization.StringSerializer" />
+      </argument>
+      <argument> <!-- type converter used on this argument -->
+        <bean class="org.hitachivantara.spring.kafka.protobuf.KafkaProtobufSerializer" />
+      </argument>
+    </bean>
+  </argument>
+  <argument type="java.lang.Boolean" value="true" /> <!-- auto flush -->
+</bean> 
+```
+
+A `BaseMessageProducer` class was added to be able to produce messages just by defining a bean for that class. That way the usage of 
```

Comment:
I like the idea of showing the simple examples first. The `BaseMessageProducer` helps here but I believe organizing things with the full communication, first simple and then complex, instead of organizing by Producer and Consumer, with each simple and complex, would help overall understanding.